Only download the file if it does not already exist

michaguenther opened the pull request:
<!-- 🚨 We recommend pull requests be filed from a non-master branch on a repository fork (e.g. <username>:fix-xxx). 🚨 -->

## Related Issues/PRs

<!--
Please reference any related feature requests, issues, or PRs here. For example, `#123`. To automatically close the referenced items when this PR is merged, please use a closing keyword (close, fix, or resolve). For example, `Close #123`. See https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue for more information.
-->
#3123

## What changes are proposed in this pull request?

Before downloading a file form the s3 bucket it is useful to check if it already exists. This saves e.g a lot of time if the same model is loaded twice with load_model.

## How is this patch tested?

<!--
If you're unsure about what to test, where to add tests, or how to run tests, please feel free to ask. We'd be happy to help.
-->
I have tested the code manually. Since the change is minimalistic and the reliance depends on os.path.exists I think it is not necessary to write a pytest function for this change.

<!--
Please describe how you confirmed the proposed feature/bug-fix/change works here. For example, if you fixed an MLflow client API, you could attach the code that didn't work prior to the fix but works now, or if you added a new feature on MLflow UI, you could attach a video that demonstrates the feature.
-->

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

## Release Notes

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

michaguenther committed:
Only download the file if it does not already exist

Before downloading a file it is useful to check if it already exists. This saves e.g a lot of time if the same model is loaded with load_model.

Signed-off-by: michaguenther <58656683+michaguenther@users.noreply.github.com>

harupy commented on mlflow/store/artifact/s3_artifact_repo.py:
```diff
@@ -190,10 +190,11 @@ def _verify_listed_object_contains_artifact_path_prefix(listed_object_path, arti
-        (bucket, s3_root_path) = data.parse_s3_uri(self.artifact_uri)
-        s3_full_path = posixpath.join(s3_root_path, remote_file_path)
-        s3_client = self._get_s3_client()
-        s3_client.download_file(bucket, s3_full_path, local_path)
```
@michaguenther Does `s3_client.download_file` overwrite the existing file?

dbczumar replied:
@michaguenther How can we verify that the local file is up-to-date?

michaguenther replied:
@dbczumar @harupy I have further developed the check. In case a file already exists locally I get the etag, hash of the file in the s3 bucket (without downloading the file in the first place). Then I compare it with the local hash and skip the download if the hashes match. The hash comparison was adopted from [stackoverflow](https://stackoverflow.com/questions/1775816/how-to-get-the-md5sum-of-a-file-on-amazons-s3).

dbczumar replied:
@michaguenther Thanks! Is there an S3 client API for computing this ETag? Alternatively, does AWS provide official code for calculating it? Reason I ask is that the etag doesn't always seem to be an md5 hash: https://stackoverflow.com/a/41735844/11952869.

michaguenther replied:
@dbczumar There seems to be a [new api](https://aws.amazon.com/blogs/aws/new-additional-checksum-algorithms-for-amazon-s3/), but this has to be activated in the aws settings. 
The code be then further optimised with the following [method](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.get_object_attributes).

I am not sure if your described issue can be resolved at the moment. 
Overall the worst thing that happens with the current code is that the file gets downloaded although it would not have been necessary but for many case we manage to skip the download.

michaguenther committed:
Check etag in order to decide about redownloading

In case a file was already downloaded before, compare the hash of the s3 server with the local file in oder to decide about skipping the download.

Signed-off-by: michaguenther <58656683+michaguenther@users.noreply.github.com>

dbczumar commented on mlflow/store/artifact/s3_artifact_repo.py:
```diff
@@ -188,13 +189,48 @@ def _verify_listed_object_contains_artifact_path_prefix(listed_object_path, arti
+            for data in iter(lambda: f.read(chunk_size), b''):
+                md5s.append(hashlib.md5(data).digest())
+        m = hashlib.md5(b"".join(md5s))
+        return '{}-{}'.format(m.hexdigest(), len(md5s))
```
`len(md5s)` seems to depend on `chunk_size`, but we can't know the actual chunk size that was used to upload the file. Accordingly, it seems like this checksum won't be valid for multi-part uploads.

michaguenther committed:
Merge branch 'mlflow:master' into master

michaguenther committed:
Merge branch 'mlflow:master' into master

mlflow-automation commented:
<!-- documentation preview -->

### Documentation preview will be available [here](https://output.circle-artifacts.com/output/job/0fb25117-5420-474c-b46b-53f243ad8ae9/artifacts/0/docs/build/html/index.html).

<details>
<summary>Notes</summary>

- Ignore this comment if this PR does not change the documentation.
- It takes a few minutes for the preview to be available.
- The preview is updated on every commit to this PR.
- Job URL: https://circleci.com/gh/mlflow/mlflow/34098
- Updated at: 2022-10-24 13:11:19.452261

</details>

ArVar commented:
Is anyone still working on that? This would be a great feature!

michaguenther commented:
> Is anyone still working on that? This would be a great feature!

@ArVar The suggested feature worked for me like a charm, but was unfortunately not approved. I guess not recommended but you can always copy this change into your local mlflow source files.
